File: backend/src/core/gamedata/types/gacha.rs
use serde::{Deserialize, Serialize};
use std::collections::HashMap;

use super::serde_helpers::deserialize_fb_map;

// ============================================================================
// Nested Structs
// ============================================================================

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct GachaPoolClient {
    #[serde(alias = "GachaPoolId")]
    pub gacha_pool_id: String,
    #[serde(alias = "GachaIndex")]
    pub gacha_index: i32,
    #[serde(alias = "OpenTime")]
    pub open_time: i64,
    #[serde(alias = "EndTime")]
    pub end_time: i64,
    #[serde(alias = "GachaPoolName")]
    pub gacha_pool_name: String,
    #[serde(alias = "GachaPoolSummary")]
    pub gacha_pool_summary: String,
    #[serde(alias = "GachaPoolDetail")]
    pub gacha_pool_detail: Option<String>,
    #[serde(alias = "Guarantee5Avail")]
    pub guarantee5_avail: i32,
    #[serde(alias = "Guarantee5Count")]
    pub guarantee5_count: i32,
    #[serde(alias = "GachaRuleType")]
    pub gacha_rule_type: String,
    #[serde(alias = "LMTGSID")]
    pub lmtgsid: Option<String>,
    #[serde(alias = "CDPrimColor")]
    pub cd_prim_color: Option<String>,
    #[serde(alias = "CDSecColor")]
    pub cd_sec_color: Option<String>,
    #[serde(alias = "LimitParam")]
    pub limit_param: Option<serde_json::Value>,
    #[serde(alias = "LinkageParam")]
    pub linkage_param: Option<serde_json::Value>,
    #[serde(alias = "LinkageRuleId")]
    pub linkage_rule_id: Option<String>,
    #[serde(alias = "DynMeta")]
    pub dyn_meta: Option<serde_json::Value>,
    #[serde(alias = "FreeBackColor")]
    pub free_back_color: Option<String>,
    #[serde(alias = "GuaranteeName")]
    pub guarantee_name: Option<String>,
}

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct NewbeeGachaPoolClient {
    #[serde(alias = "GachaPoolId")]
    pub gacha_pool_id: String,
    #[serde(alias = "GachaIndex")]
    pub gacha_index: i32,
    #[serde(alias = "GachaPoolName")]
    pub gacha_pool_name: String,
    #[serde(alias = "GachaPoolDetail")]
    pub gacha_pool_detail: Option<String>,
    #[serde(alias = "GachaPrice")]
    pub gacha_price: i32,
    #[serde(alias = "GachaTimes")]
    pub gacha_times: i32,
    #[serde(alias = "GachaOffset")]
    pub gacha_offset: Option<String>,
}

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct GachaTag {
    #[serde(alias = "TagId")]
    pub tag_id: i32,
    #[serde(alias = "TagName")]
    pub tag_name: String,
    #[serde(alias = "TagGroup")]
    pub tag_group: i32,
}

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct RecruitTimeEntry {
    #[serde(alias = "RecruitPrice")]
    pub recruit_price: i32,
}

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct RecruitPool {
    #[serde(alias = "RecruitTimeTable")]
    pub recruit_time_table: Vec<RecruitTimeEntry>,
    // Add other fields as needed
}

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct RecruitRarityEntry {
    #[serde(alias = "RarityStart")]
    pub rarity_start: i32,
    #[serde(alias = "RarityEnd")]
    pub rarity_end: i32,
}

// ============================================================================
// Container Type
// ============================================================================

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct GachaData {
    pub gacha_pool_client: Vec<GachaPoolClient>,
    pub newbee_gacha_pool_client: Vec<NewbeeGachaPoolClient>,
    pub special_recruit_pool: Vec<serde_json::Value>,
    pub gacha_tags: Vec<GachaTag>,
    pub recruit_pool: RecruitPool,
    pub potential_material_converter: serde_json::Value,
    pub classic_potential_material_converter: serde_json::Value,
    pub recruit_rarity_table: HashMap<i32, RecruitRarityEntry>,
    pub special_tag_rarity_table: HashMap<i32, Vec<i32>>,
    pub recruit_detail: String,
    pub show_gacha_log_entry: bool,
    pub carousel: Vec<serde_json::Value>,
    pub free_gacha: Vec<serde_json::Value>,
    pub limit_ten_gacha_item: Vec<serde_json::Value>,
    pub linkage_ten_gacha_item: Vec<serde_json::Value>,
    pub normal_gacha_item: Vec<serde_json::Value>,
    pub fes_gacha_pool_relate_item: HashMap<String, FesGachaPoolRelateEntry>,
    pub dic_recruit6_star_hint: HashMap<String, String>,
    pub special_gacha_percent_dict: HashMap<i32, f64>,
}

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct FesGachaPoolRelateEntry {
    #[serde(alias = "RarityRank5ItemId")]
    pub rarity_rank5_item_id: String,
    #[serde(alias = "RarityRank6ItemId")]
    pub rarity_rank6_item_id: String,
}

// ============================================================================
// Table File Wrapper
// ============================================================================

#[derive(Debug, Clone, Default, Deserialize)]
#[serde(rename_all = "PascalCase")]
pub struct GachaTableFile {
    pub gacha_pool_client: Vec<GachaPoolClient>,
    pub newbee_gacha_pool_client: Vec<NewbeeGachaPoolClient>,
    #[serde(default)]
    pub special_recruit_pool: Vec<serde_json::Value>,
    pub gacha_tags: Vec<GachaTag>,
    pub recruit_pool: RecruitPool,
    #[serde(default)]
    pub potential_material_converter: serde_json::Value,
    #[serde(default)]
    pub classic_potential_material_converter: serde_json::Value,
    #[serde(deserialize_with = "deserialize_fb_map", default)]
    pub recruit_rarity_table: HashMap<i32, RecruitRarityEntry>,
    #[serde(deserialize_with = "deserialize_fb_map", default)]
    pub special_tag_rarity_table: HashMap<i32, Vec<i32>>,
    #[serde(default)]
    pub recruit_detail: String,
    #[serde(default)]
    pub show_gacha_log_entry: bool,
    #[serde(default)]
    pub carousel: Vec<serde_json::Value>,
    #[serde(default)]
    pub free_gacha: Vec<serde_json::Value>,
    #[serde(default)]
    pub limit_ten_gacha_item: Vec<serde_json::Value>,
    #[serde(default)]
    pub linkage_ten_gacha_item: Vec<serde_json::Value>,
    #[serde(default)]
    pub normal_gacha_item: Vec<serde_json::Value>,
    #[serde(deserialize_with = "deserialize_fb_map", default)]
    pub fes_gacha_pool_relate_item: HashMap<String, FesGachaPoolRelateEntry>,
    #[serde(deserialize_with = "deserialize_fb_map", default)]
    pub dic_recruit6_star_hint: HashMap<String, String>,
    #[serde(deserialize_with = "deserialize_fb_map", default)]
    pub special_gacha_percent_dict: HashMap<i32, f64>,
}

File: frontend/src/types/api/impl/gacha.ts
// Gacha types

// ============================================
// v3 Gacha Record Types (raw DB row — still used by community + settings pages)
// ============================================

/** v3 gacha record from database */
export interface GachaRecord {
    id: number;
    user_id: string;
    char_id: string;
    pool_id: string;
    rarity: number;
    pull_timestamp: number;
    pool_name: string | null;
    gacha_type: string | null;
    created_at: string;
}

/** v3 user gacha statistics */
export interface GachaStats {
    user_id: string;
    total_pulls: number | null;
    six_star_count: number | null;
    five_star_count: number | null;
    four_star_count: number | null;
    first_pull: number | null;
    last_pull: number | null;
}

/** v3 global gacha statistics (camelCase from service layer). Used by community page. */
export interface GachaGlobalStats {
    totalPulls: number;
    totalUsers: number;
    sixStarRate: number;
    fiveStarRate: number;
}

/** v3 gacha fetch result (camelCase from service layer). */
export interface GachaFetchResult {
    totalFetched: number;
    newRecords: number;
}

// ============================================
// Grouped GachaRecords (limited/regular/special) — used by history page
// ============================================

export type GachaType = "limited" | "regular" | "special";

export interface GachaItem {
    charId: string;
    charName: string;
    star: string;
    color: string;
    poolId: string;
    poolName: string;
    typeName: string;
    at: number;
    atStr: string;
}

export interface GachaTypeRecords {
    gacha_type: GachaType;
    records: GachaItem[];
    total: number;
}

export interface GachaRecords {
    limited: GachaTypeRecords;
    regular: GachaTypeRecords;
    special: GachaTypeRecords;
}

// ============================================
// History / Settings
// ============================================

/** Individual pull entry from stored history */
export interface GachaRecordEntry {
    id: string;
    charId: string;
    charName: string;
    rarity: number;
    poolId: string;
    poolName: string;
    gachaType: string;
    pullTimestamp: number;
    pullTimestampStr: string | null;
}

/** Pagination metadata for gacha history */
export interface GachaPaginationInfo {
    limit: number;
    offset: number;
    total: number;
    hasMore: boolean;
}

export interface DateRange {
    from: number | null;
    to: number | null;
}

export interface HistoryFilters {
    rarity: number | null;
    gachaType: string | null;
    charId: string | null;
    dateRange: DateRange | null;
}

export interface GachaHistoryResponse {
    records: GachaRecordEntry[];
    pagination: GachaPaginationInfo;
    filtersApplied: HistoryFilters;
}

export interface GachaHistoryParams {
    limit?: number;
    offset?: number;
    rarity?: number;
    gachaType?: GachaType;
    charId?: string;
    from?: number;
    to?: number;
    order?: "asc" | "desc";
}

export interface GachaSettings {
    user_id: string;
    store_records: boolean;
    share_anonymous_stats: boolean;
    total_pulls: number;
    six_star_count: number;
    five_star_count: number;
    last_sync_at: string | null;
}

// ============================================
// Gacha Pool Types (Game Client Data)
// ============================================

export interface GachaPoolClient {
    gachaPoolId: string;
    gachaIndex: number;
    openTime: number;
    endTime: number;
    gachaPoolName: string;
    gachaPoolSummary: string;
    gachaPoolDetail: string | null;
    guarantee5Avail: number;
    guarantee5Count: number;
    gachaRuleType: string;
    lmtgsid: string | null;
    cdPrimColor: string | null;
    cdSecColor: string | null;
    limitParam: unknown | null;
    linkageParam: unknown | null;
    linkageRuleId: string | null;
    dynMeta: unknown | null;
    freeBackColor: string | null;
    guaranteeName: string | null;
}

export interface NewbeeGachaPoolClient {
    gachaPoolId: string;
    gachaIndex: number;
    gachaPoolName: string;
    gachaPoolDetail: string | null;
    gachaPrice: number;
    gachaTimes: number;
    gachaOffset: string | null;
}

export interface GachaTag {
    tagId: number;
    tagName: string;
    tagGroup: number;
}

export interface RecruitTimeEntry {
    recruitPrice: number;
}

export interface RecruitPool {
    recruitTimeTable: RecruitTimeEntry[];
}

export interface RecruitRarityEntry {
    rarityStart: number;
    rarityEnd: number;
}

export interface FesGachaPoolRelateEntry {
    rarityRank5ItemId: string;
    rarityRank6ItemId: string;
}

export interface GachaData {
    gachaPoolClient: GachaPoolClient[];
    newbeeGachaPoolClient: NewbeeGachaPoolClient[];
    specialRecruitPool: unknown[];
    gachaTags: GachaTag[];
    recruitPool: RecruitPool;
    potentialMaterialConverter: unknown;
    classicPotentialMaterialConverter: unknown;
    recruitRarityTable: Record<number, RecruitRarityEntry>;
    specialTagRarityTable: Record<number, number[]>;
    recruitDetail: string;
    showGachaLogEntry: boolean;
    carousel: unknown[];
    freeGacha: unknown[];
    limitTenGachaItem: unknown[];
    linkageTenGachaItem: unknown[];
    normalGachaItem: unknown[];
    fesGachaPoolRelateItem: Record<string, FesGachaPoolRelateEntry>;
    dicRecruit6StarHint: Record<string, string>;
    specialGachaPercentDict: Record<number, number>;
}

/** Response from /static/gacha/pools endpoint */
export interface GachaPoolsResponse {
    pools: GachaPoolClient[];
    newbeePools: NewbeeGachaPoolClient[];
    total: number;
}

export interface GachaTableFile {
    gachaPoolClient: GachaPoolClient[];
    newbeeGachaPoolClient: NewbeeGachaPoolClient[];
    specialRecruitPool: unknown[];
    gachaTags: GachaTag[];
    recruitPool: RecruitPool;
    potentialMaterialConverter: unknown;
    classicPotentialMaterialConverter: unknown;
    recruitRarityTable: Record<number, RecruitRarityEntry>;
    specialTagRarityTable: Record<number, number[]>;
    recruitDetail: string;
    showGachaLogEntry: boolean;
    carousel: unknown[];
    freeGacha: unknown[];
    limitTenGachaItem: unknown[];
    linkageTenGachaItem: unknown[];
    normalGachaItem: unknown[];
    fesGachaPoolRelateItem: Record<string, FesGachaPoolRelateEntry>;
    dicRecruit6StarHint: Record<string, string>;
    specialGachaPercentDict: Record<number, number>;
}
